### PR TITLE
Store irrigation as volume or time or hydrometer indication (fixes #291)

### LINKED ISSUE #291: Store irrigation as volume or time or hydrometer indication
Currently the user enters irrigations he has applied as volume of water, i.e. he says "on 2020-04-10 10:04, I watered the field with 10 m³". But very often he doesn't know directly how much water he threw in, and he must calculate it from the duration of irrigation or from the indication of a hydrometer. It's just a multiplication or a subtraction, but we want the system to do it rather than the user.

So, each time the user wants to enter an applied irrigation, he should be presented with this list of options:

- I want to specify the volume of water
- I want to specify the duration of irrigation
- I want to register the hydrometer readings

The default option should be the same as the one used for the previous applied irrigation (or no default if no previous applied irrigation).

When the user selects a different option, the fields he needs to enter should change accordingly.

In all three cases, the user must specify the date and time at the end of the irrigation.

If the user selects **duration of irrigation**, there must be two more fields:
1. Duration of irrigation, in hours and minutes
2. Flow rate estimated by inspection, in m³/h. The default for this field must be the one used for the previous applied irrigation (empty if no previous applied irrigation).

If the user selects **hydrometer readings**, there must the following four items (in addition to the date and time at the end of the irrigation):
1. Percentage of water that corresponds to this field (%). The default must be the same as the previous applied irrigation, or 100 if there's no previous applied irrigation.
2. Previous hydrometer reading (this is not a form field, it's just information for the user).
3. Hydrometer reading at start of irrigation. Default is the same as the previous hydrometer reading, empty if no previous hydrometer reading.
4. Hydrometer reading at end of irrigation. No default.

Hydrometer readings are actually three numbers: e.g. 382 m³, 9 ×0.1, 7 ×0.01. This means 382.97 m³ (and I think it's a good idea to store it as a `FloatField` or `DecimalField`), but hydrometers display it as m³, ×0.1, ×.01, so we are going to use something similar. (The × is the multiplication sign, 0xD7.)

### PR DESCRIPTION
**Pull request by AbdullahKady**

### CODE REVIEW

**Comment by AbdullahKady** [comment]
@aptiko I'll address the review comments here, so we can keep any further discussion on this PR:

As for the function being too big:
I get the "3 different parts", I did split it but not all the way through. I get it from a clean code point of view, but that's not always correct since the function is pretty straight forward (the conditions/effects are way too simple, that they could be replaced for a mapper for all it's worth). I think splitting each one of the try/except blocks will actually be a bit more unreadable. Let me know what you think.

**Comment by AbdullahKady** [comment]
Finally, for default watering:

As for the default when no water volume is present, I didn't completely remove it, I just moved it around since it didn't seem in the correct place (the model was attaching the view's error message).
But as you said, it makes sense to have it as its own separate type indeed. Will add it and provide a data migration to move all applied_irrigations with 0 volume to the newly added irrigation type.

**Comment by aptiko** [comment]
OK, regarding `specified`/`supplied`, my error. I was confused. I thought that "supplied water amount" means "amount of water that has been supplied to the (agri)field". But you are using it as a synonym for "specified". Keep whatever names you like best.

**Comment by aptiko** [comment]
Let's have some fun with the long function.

First of all, I find your split in two more unreadable than the original which was just a single function. So let's use your original as a base.

This seems more readable to me:

```
    def get_latest_applied_irrigation_defaults(self):
        """
        Returns a dict of all default values from the history of AppliedIrrigations
        per this field. Note that some dict keys won't exist if no previous values.
        """
        result = {}
        result.update(self._get_applied_irrigation_default_type())
        result.update(self._get_applied_irrigation_defaults_for_volume())
        result.update(self._get_applied_irrigation_defaults_for_duration())
        result.update(self._get_applied_irrigation_defaults_for_hydrometer())
        return result

    def _get_applied_irrigation_default_type(self):
        try:
            return {
                "irrigation_type": self.appliedirrigation_set.latest().irrigation_type
            }
        except AppliedIrrigation.DoesNotExist:
            return {}

    def _get_applied_irrigation_defaults_for_volume(self):
        try:
            return {
                "supplied_water_volume": self.appliedirrigation_set.filter(
                    irrigation_type="VOLUME_OF_WATER"
                )
                .latest()
                .supplied_water_volume
            }
        except AppliedIrrigation.DoesNotExist:
            return {}

    def _get_applied_irrigation_defaults_for_duration(self):
        try:
            duration_irr = self.appliedirrigation_set.filter(
                irrigation_type="DURATION_OF_IRRIGATION"
            ).latest()
            return {
                "supplied_duration": duration_irr.supplied_duration,
                "supplied_flow_rate": duration_irr.supplied_flow_rate,
            }
        except AppliedIrrigation.DoesNotExist:
            return {}

    def _get_applied_irrigation_defaults_for_hydrometer(self):
        try:
            hydro_irr = self.appliedirrigation_set.filter(
                irrigation_type="HYDROMETER_READINGS"
            ).latest()
            return {
                "hydrometer_water_percentage": hydro_irr.hydrometer_water_percentage,
                "hydrometer_reading_start": hydro_irr.hydrometer_reading_end,
            }
        except AppliedIrrigation.DoesNotExist:
            return {}
```

However, I still find the "subfunctions", so to speak, too complicated. However it's hard to make them better. Maybe by taking the body of the "try" into a separate function, but it would be hard to find a name for that function. However, trying to find a good name for that function might lead us to some good refactoring.

BTW, I think that a better name for the top-level function would be `get_applied_irrigation_defaults()`. I think that the word "latest" isn't clarifying anything.

Now, you find a single function more readable? If you find version A more readable and I find version B more readable, what do we do? Maybe we should have some coding specifications? My coding specifications are Robert C. Martin's "Clean Code". But I could argue that since R. M. says a function should be no more than 5 lines, and the shortest and more readable function of those above is already 6 lines, not to mention the lower level functions, then I'm not following R. M. anyway except as a guideline which I've already broken. I draw the line differently than you draw yours. So go figure.

But speaking of R. M., I take it a step further. He says classes should be small, and that they should be even smaller than that. The `Agrifield` class is already too large (and the way I've split some functionality away into `agrifield.py` is far from perfect). If I read him correctly, he'd say that what we are trying to do here is a unit of functionality that could be split into another class.

Bottom line: Do whatever you like. If you think your original version was the most readable, use that. If you think your second version (which I find less readable) is more readable, use that. If you find my version above more readable, use that. If you are intrigued by the exercise "readability-on-steroids/Robert-Martin-compliance", go ahead and spend one or two hours on it.

**Comment by aptiko** [comment]
Final comments: There's a PEP for the docstrings, please use it. Then fix the errors CI found, then we're ready to merge.

**Comment by AbdullahKady** [comment]
Regarding the function, this was the way I had it setup in the first place (4 different functions, and one aggregating them), matter of fact, I actually had a basic decorator that returns an empty dict on `AppliedIrrigation.DoesNotExist`, but when I finished the code I felt like it was a bit too much splitting. Thinking in the future if I come to read this `models` file, I shouldn't expect 4 private functions that are almost doing the same thing.
To be honest, the rule of 5-loc is pretty unrealistic in my opinion. I like to read discussions aside from theoretical aspects (reddit/stackexchange) regarding clean code. For instance this entire thing can be replaced with dict/mapper, and thus the actual code will be a single line of code, however to me that's much more unreadable. So as far as the function doesn't have complex/nested flows/paths, I felt like it was appropriate.

Please note that I really enjoy these discussions as I find them very beneficial, that's the only reason I'm actually listing my reasoning/thought-process.

**Comment by AbdullahKady** [comment]
As for the pep8 docstring, I'm not really sure what you mean?

When I searched before, I found that pep8 doesn't really enforce much about the actual docstring, but rather the basic stuff. Mainly people choose to follow standards such as google's (specially for adding return/argument types, as well as examples).
So according to [this](https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings), the only thing I ended up changing is adding an extra blank line after the "summary".

**Comment by aptiko** [comment]
Yes, that was what I meant, that docstrings must have a one-line "summary". I think it's more common for that summary to be on the same line as the opening quotes, although PEP 257 allows it to be in the next line.

This is probably ready to merge, but, to be sure, I'll do some manual testing on this later. I'll probably also translate the strings.

**Comment by AbdullahKady** [comment]
> This is probably ready to merge, but, to be sure, I'll do some manual testing on this later. I'll probably also translate the strings.

Yeah I totally forgot, I wanted to `makemessages` and leave them up to you for translation. Nice catch

**Comment by aptiko** [comment]
The UI needs some improvements.

# 1 

Screenshot from 2020-04-27 11-53-09

When you click on a past irrigation, it shows you an edit form which does not work as it should (it doesn't hide and show fields as needed, it seems to have all fields visible at the same time—hydrometer readings, flow rate, duration, amount).

# 2

Instead of a "Type of irrigation" dropdown, use an (untitled) radio box, and phrase the three options like this:

- I want to specify the volume of water
- I want to specify the duration of irrigation
- I want to register the hydrometer readings

# 3

We used to have a date/time selection widget that really sucked. Whether we keep the same one or a better one, we do need a widget and not expect the user to fill-in the date/time in iso8601 format. (I think that the time only needs to be specified accurate to the hour.)

# 4

Use this phrasing for the various fields:

- Duration of irrigation
- Percentage of water that corresponds to this field
- Hydrometer reading at start of irrigation
- Hydrometer reading at end of irrigation

# 5

Add the "previous hydrometer reading" as I've specified it in the description of #291.

**Comment by AbdullahKady** [comment]
> The UI needs some improvements.
> 
> # 1
> Screenshot from 2020-04-27 11-53-09
> 
> When you click on a past irrigation, it shows you an edit form which does not work as it should (it doesn't hide and show fields as needed, it seems to have all fields visible at the same time—hydrometer readings, flow rate, duration, amount).
> 
> # 2
> Instead of a "Type of irrigation" dropdown, use an (untitled) radio box, and phrase the three options like this:
> 
> * I want to specify the volume of water
> * I want to specify the duration of irrigation
> * I want to register the hydrometer readings
> 
> # 3
> We used to have a date/time selection widget that really sucked. Whether we keep the same one or a better one, we do need a widget and not expect the user to fill-in the date/time in iso8601 format. (I think that the time only needs to be specified accurate to the hour.)
> 
> # 4
> Use this phrasing for the various fields:
> 
> * Duration of irrigation
> * Percentage of water that corresponds to this field
> * Hydrometer reading at start of irrigation
> * Hydrometer reading at end of irrigation
> 
> # 5
> Add the "previous hydrometer reading" as I've specified it in the description of #291.

1) Oops, didn't even know that irrigation(s) were editable, will work on it.
2) :+1: 
3) I actually did comment on that (I wrote down what irritated me as UX while signing up), but I forgot to send it to you. Didn't bother changing it since it was not introduced in this task. Should I do it in the same task anyway?
4) :+1: 
5) Missed that, sorry. How/Where would you like them to be displayed? I'm not very good with rendering forms, so if you can point me in any direction it would be great!

**Comment by aptiko** [comment]
On 5, eh, omit it. Let's see what the agronomists will say.

On 3, no, you don't need to change it, but I think it disappeared. I don't see any way to bring up the widget. AFAICS it used to work with on-enter or something.

**Comment by aptiko** [comment]
Maybe it wasn't you; maybe it disappeared when I renamed the model and fields.

**Comment by AbdullahKady** [comment]
1) So the forms have different DOM structure, I could've done a while loop to keep going up the chain [here](), but I felt like it might be better that if someone adds a new form and uses this function without actually testing (hopefully that never happens anyway), it would break by showing the wrong field(s) rather than crashing the browser with an infinite while loop.

2) After some search, I didn't find a 'clean' way to override the display values of the choice, except for overriding them in the model which didn't seem right to me; because the 'I want to ...' display wouldn't make sense in say admin or any other view for that matter.

3) Fixed, name change indeed.

4) Done, I generated the `.po` files so you can translate them.

**Comment by aptiko** [comment]
Two problems:
1. If a field has no irrigations in the past, and I visit the irrigations page, all the fields are visible. I think that "volume" should be the default.
2. Could you rename migration 0030 to something more descriptive?

**Comment by AbdullahKady** [comment]
> Two problems:
> 
> 1. If a field has no irrigations in the past, and I visit the irrigations page, all the fields are visible. I think that "volume" should be the default.
> 2. Could you rename migration 0030 to something more descriptive?

Done, please check
